pinot-common/src/main/java/org/apache/pinot/common/utils/RegexpPatternConverterUtils.java

```diff
@@ -64,27 +70,54 @@ public static String likeToRegexpLike(String likePattern) {
         break;
     }
 
-    String escaped = escapeMetaCharacters(likePattern.substring(start, end));
-    StringBuilder sb = new StringBuilder(escaped.length() + 2);
+    likePattern = likePattern.substring(start, end);
+    StringBuilder sb = new StringBuilder();
     sb.append(prefix);
-    sb.append(escaped);
-    sb.append(suffix);
 
+    // handling SQL wildcards by replacing them with corresponding regex equivalents
+    // we ignore them if the SQL wildcards are escaped
     int i = 0;
-    while (i < sb.length()) {
-      char c = sb.charAt(i);
+    boolean isPrevCharBackSlash = false;
+    while (i < likePattern.length()) {
+      char c = likePattern.charAt(i);
       if (c == '_') {
-        sb.replace(i, i + 1, ".");
+        sb.append(isPrevCharBackSlash ? c : ".");
       } else if (c == '%') {
-        sb.replace(i, i + 1, ".*");
-        i++;
+        sb.append(isPrevCharBackSlash ? c : ".*");
+      } else if (indexOf(REGEXP_METACHARACTERS, c) >= 0) {
+        sb.append(BACK_SLASH).append(c);
+      } else {
+        if (isPrevCharBackSlash) {
+          // this means the previous character is a \
+          // but it was not used for escaping SQL wildcards
+          // so let's escape this \ in the output
+          // this case is separately handled outside of the meta characters list
+          sb.append(BACK_SLASH);
+        }
+        sb.append(c);
       }
-      i++;
+      isPrevCharBackSlash = (c == BACK_SLASH);
+      ++i;
+    }
+
+    // handle trailing \
+    if (isPrevCharBackSlash) {
+      sb.append(BACK_SLASH);
     }
 
+    sb.append(suffix);
     return sb.toString();
   }
 
+  private static int indexOf(char[] arr, char c) {
```

Comment:
why re-imple indexOf rather than using existing lib such as

https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/IterableUtils.html#find-java.lang.Iterable-org.apache.commons.collections4.Predicate-

Reply:
Tried searching for it before -- didn't find one in an already included package.

But found Guava's Chars.indexOf which does this -- replaced it with that.